extensions/emmet/src/abbreviationActions.ts

```diff
@@ -37,45 +43,166 @@ export function wrapWithAbbreviation(args: any) {
+			if (previewMade) {
+				returnPromise = revertPreview(editor, rangesToReplace).then(() => { return false; });
+			}
+			return returnPromise;
+		}
 
 		let extractedResults = helper.extractAbbreviationFromText(inputAbbreviation);
 		if (!extractedResults) {
-			return false;
+			return Promise.resolve(previewMade);
 		}
 		let { abbreviation, filter } = extractedResults;
 
-		let expandAbbrList: ExpandAbbreviationInput[] = [];
+		expandAbbrList = [];
 
-		editor.selections.forEach(selection => {
-			let rangeToReplace: vscode.Range = selection.isReversed ? new vscode.Range(selection.active, selection.anchor) : selection;
-			if (rangeToReplace.isEmpty) {
-				let { active } = selection;
-				let currentNode = getNode(rootNode, active, true);
-				if (currentNode && (currentNode.start.line === active.line || currentNode.end.line === active.line)) {
-					rangeToReplace = new vscode.Range(currentNode.start, currentNode.end);
-				} else {
-					rangeToReplace = new vscode.Range(rangeToReplace.start.line, 0, rangeToReplace.start.line, editor.document.lineAt(rangeToReplace.start.line).text.length);
-				}
+		// we need to apply the previewchanges and get the new ranges
+		let revertPromise: Thenable<any> = Promise.resolve();
+		if (definitive) {
+			if (previewMade) {
+				revertPromise = revertPreview(editor, rangesToReplace);
 			}
+			rangesToReplace.forEach(rangesAndContent => {
+				let rangeToReplace = rangesAndContent.originalRange;
+				let textToWrap = rangeToReplace.isSingleLine ? ['$TM_SELECTED_TEXT'] : ['\n\t$TM_SELECTED_TEXT\n'];
+				expandAbbrList.push({ syntax: syntax || '', abbreviation, rangeToReplace, textToWrap, filter });
+			});
+			return revertPromise.then(() => {
+				return expandAbbreviationInRange(editor, expandAbbrList, true).then(() => { return Promise.resolve(); });
+			});
+		} else {
+			rangesToReplace.forEach(rangesAndContent => {
+				let match = rangesAndContent.content.match(/\n[\s]*/g);
+				let textToWrap = match ? ['\n\t' + rangesAndContent.content.split(match[match.length - 1]).join('\n\t') + '\n'] : [rangesAndContent.content];
+				expandAbbrList.push({ syntax: syntax || '', abbreviation, rangeToReplace: rangesAndContent.originalRange, textToWrap, filter });
+			});
+		}
 
-			const firstLineOfSelection = editor.document.lineAt(rangeToReplace.start).text.substr(rangeToReplace.start.character);
-			const matches = firstLineOfSelection.match(/^(\s*)/);
-			const preceedingWhiteSpace = matches ? matches[1].length : 0;
-
-			rangeToReplace = new vscode.Range(rangeToReplace.start.line, rangeToReplace.start.character + preceedingWhiteSpace, rangeToReplace.end.line, rangeToReplace.end.character);
-			let textToWrap = rangeToReplace.isSingleLine ? ['$TM_SELECTED_TEXT'] : ['\n\t$TM_SELECTED_TEXT\n'];
-			expandAbbrList.push({ syntax, abbreviation, rangeToReplace, textToWrap, filter });
+		return Promise.resolve().then(() => {
+			return applyPreview(editor, expandAbbrList, rangesToReplace).then(ranges => {
```

Comment:
Why return the ranges here from `applyPreview` and the update `rangesToReplace`?
`rangesToReplace` is accessible from inside `applyPreview` and can be updated from there as well.
Just like you did in `revertPreview`